### PR TITLE
feat(behavior_path_planner): update pull_over module

### PR DESCRIPTION
**Pull request by kosuke55**
## Description

Updated pull_over_module. Added the following features

- goal search. find free space for parking using occupancy grid.
  - https://user-images.githubusercontent.com/39142679/178033628-bec1bbc7-3b27-47b1-b50b-55f855e2e399.mp4
- multiple parking method considering how large is the space
  - shift parking
    - https://user-images.githubusercontent.com/39142679/178034101-4dc61a33-bc49-41a0-a9a8-755cce53cbc6.mp4
  - arc forward parking 
    - https://user-images.githubusercontent.com/39142679/178034128-4754c401-8aff-4745-b69a-4a69ca29ce4b.mp4
  - arc backward parking 
    - https://user-images.githubusercontent.com/39142679/178034280-4b6754fe-3981-4aee-b5e0-970f34563c6d.mp4
- wait until space in available
https://user-images.githubusercontent.com/39142679/178034292-709c47bf-af82-44af-a11d-850bb3e061f8.mp4

## Tests performed

psim, scenario test, real vehicle.

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].
- [ ] The PR has been properly tested.
- [ ] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

**Comment by yukkysaito** [comment]
Great. I'd like to experiment in test fields before merge.

**Comment by kosuke55** [comment]
Merge https://github.com/autowarefoundation/autoware.universe/pull/901 first.

**Comment by yukkysaito** [comment]
PR is too big. Can you separate this for utils

File: planning/behavior_path_planner/config/pull_over/pull_over.param.yaml [hunk 1]
```diff
@@ -1,26 +1,54 @@
...
+      before_pull_over_straight_distance: 5.0
+      # parallel parking path
+      after_forward_parking_straight_distance: 2.0
+      after_backward_parking_straight_distance: 2.0
+      forward_parking_velocity: 0.3
+      backward_parking_velocity: -0.3
+      arc_path_interval: 1.0
+      # hazard. Not used now.
+      double hazard_on_threshold_dis: 1.0
+      double hazard_on_threshold_vel: 0.5
```
Comment (satoshi-ota) [inline planning/behavior_path_planner/config/pull_over/pull_over.param.yaml:43]: no need `double` ?

**Comment by kosuke55** [comment]
(WIP) separate this PR to
- https://github.com/autowarefoundation/autoware.universe/pull/1257
-

File: planning/behavior_path_planner/src/scene_module/pull_over/pull_over_module.cpp [hunk 2]
```diff
@@ -96,115 +188,424 @@ bool PullOverModule::isExecutionRequested() const
...
+    } else {
+      output.turn_signal_info.turn_signal.command = turn_info.first.command;
+      output.turn_signal_info.signal_distance = turn_info.second;
+    }
+  }
+
+  // For experimets
+  if (parameters_.print_debug_info) {
+    printParkingPositionError();
   }
```
Comment (rej55) [inline planning/behavior_path_planner/src/scene_module/pull_over/pull_over_module.cpp:579]:
```
updateRTCStatus(distance);
```

File: planning/behavior_path_planner/src/scene_module/pull_over/pull_over_module.cpp [hunk 3]
```diff
@@ -96,115 +188,424 @@ bool PullOverModule::isExecutionRequested() const
...
+
+    if (*dist_to_parking_start_pose < parameters_.decide_path_distance) {
+      status_.has_decided_path = true;
+    }
+  }
+
+  // Use decided path
+  if (status_.has_decided_path) {
+    if (!status_.has_requested_approval_) {
+      waitApproval();
```
[the remark this answers is not in the document]
Comment (kosuke55) [reply planning/behavior_path_planner/src/scene_module/pull_over/pull_over_module.cpp:496]: is it OK after `waitApproval()`

[the remark this answers is not in the document]
Comment (rej55) [reply planning/behavior_path_planner/src/scene_module/pull_over/pull_over_module.cpp:496]: Yes.
`waitApproval()` just inverts the flag, so the order is not important.

https://github.com/autowarefoundation/autoware.universe/blob/488bd4212858e4127351c13b045f869046a27da7/planning/behavior_path_planner/include/behavior_path_planner/scene_module/scene_module_interface.hpp#L269

File: planning/behavior_path_planner/config/pull_over/pull_over.param.yaml (hunk 1 above)
Comment (kosuke55) [reply planning/behavior_path_planner/config/pull_over/pull_over.param.yaml]: fixed in https://github.com/autowarefoundation/autoware.universe/pull/873/commits/652e424f8a879c8f26a6c4bcd4edd724313b5420

File: planning/behavior_path_planner/src/scene_module/pull_over/pull_over_module.cpp (hunk 2 above)
Comment (kosuke55) [reply planning/behavior_path_planner/src/scene_module/pull_over/pull_over_module.cpp:585]: fix in https://github.com/autowarefoundation/autoware.universe/pull/873/commits/39d02a128edc284dd7ef1052a78f03fe257e65bf

File: planning/behavior_path_planner/src/scene_module/pull_over/pull_over_module.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (kosuke55) [reply planning/behavior_path_planner/src/scene_module/pull_over/pull_over_module.cpp:494]: fix in https://github.com/autowarefoundation/autoware.universe/pull/873/commits/39d02a128edc284dd7ef1052a78f03fe257e65bf

**Comment by kosuke55** [comment]
@yukkysaito @tkimura4 
All separated PR has been merged so squashed this PR.

**Review by yukkysaito — APPROVED** [review]
LGTM

File: map/lanelet2_extension/lib/utilities.cpp [hunk 4]
```diff
@@ -24,6 +24,7 @@
 #include <lanelet2_core/primitives/BasicRegulatoryElements.h>
 #include <lanelet2_traffic_rules/TrafficRules.h>
 #include <lanelet2_traffic_rules/TrafficRulesFactory.h>
+#include <tf2_geometry_msgs/tf2_geometry_msgs.h>
```
Comment (tkimura4) [inline map/lanelet2_extension/lib/utilities.cpp:27]: tf2_geometry_msgs.hpp ? (for humble build)

Comment (kosuke55) [reply map/lanelet2_extension/lib/utilities.cpp:29]: thanks fixed in https://github.com/autowarefoundation/autoware.universe/pull/873/commits/c1fbfc6cd7b0668cbbf2a3bf7a6ac6681bae7c7f

**Pull request merged**